aten/src/ATen/native/mps/operations/HistogramKernel.mm

```diff
@@ -0,0 +1,395 @@
+
+enum BIN_SELECTION_ALGORITHM {
+  LINEAR_INTERPOLATION,
+  LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH,
+  BINARY_SEARCH,
+};
+
+static const char* METAL_HISTOGRAM = R"HISTOGRAM_METAL(
+
+#include <metal_stdlib>
+using namespace metal;
+
+enum BIN_SELECTION_ALGORITHM {
+  LINEAR_INTERPOLATION,
+  LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH,
+  BINARY_SEARCH,
+};
+
+// Re-implementation of std::upper_bound with some modifications.
+template<typename T>
+int64_t upper_bound(constant T * arr, int64_t first, int64_t len, T val) {
+  int64_t middle;
+  int64_t half_;
+
+  while (len > 0) {
+    half_ = len >> 1;
+    middle = first + half_;
+
+    if (val < arr[middle]) {
+      len = half_;
+    } else {
+      first = ++middle;
+      len = len - half_ - 1;
+    }
+  }
+  return first;
+}
+
+// The implementation here is mostly taken from the CPU's implementation with some modifications.
+// Please see `aten/src/ATen/native/cpu/HistogramKernel.cpp` for more details.
+template<typename T>
+kernel void histogramdd(constant T  * input_            [[buffer(0)]],
+                  constant T        * weight            [[buffer(1)]],
+                  device   T        * local_out         [[buffer(2)]],
+                  constant uint     * offsets           [[buffer(3)]],
+                  constant size_t   & num_dims          [[buffer(4)]],
+                  constant T        * bin_seq           [[buffer(5)]],
+                  constant int64_t  * num_bin_edges     [[buffer(6)]],
+                  constant T        * leftmost_edge     [[buffer(7)]],
+                  constant T        * rightmost_edge    [[buffer(8)]],
+                  constant int64_t  * local_out_strides [[buffer(9)]],
+                  constant uint8_t  & algorithm         [[buffer(10)]],
+                  constant uint8_t  & has_weight        [[buffer(11)]],
+                  uint tid [[thread_position_in_grid]]) {
+
+  constexpr T eps = 8e-7;
+  bool skip_element = false;
+  int64_t hist_index = 0;
+  int64_t bin_seq_offset = 0;
+
+  for (size_t dim = 0; dim < num_dims; dim++) {
+    T element = input_[(size_t)offsets[tid * num_dims + dim]];
+
+    // Skips elements which fall outside the specified bins and NaN elements
+    // Adding an eps to the edges to eliminate precision issues that cause elements accidentally skipped,
+    // this is likely due to the minuscule implementation differences between the CPU and MPS's linspace.
+    if (!(element >= (leftmost_edge[dim] - eps) && element <= (rightmost_edge[dim] + eps))) {
+        skip_element = true;
+        break;
+    }
+    int64_t pos = -1;
+
+    if (algorithm == BIN_SELECTION_ALGORITHM::BINARY_SEARCH) {
+      pos = upper_bound(
+        bin_seq,
+        bin_seq_offset,
+        num_bin_edges[dim],
+        element
+      ) - bin_seq_offset - 1;
+    } else if (
+      algorithm == BIN_SELECTION_ALGORITHM::LINEAR_INTERPOLATION ||
+      algorithm == BIN_SELECTION_ALGORITHM::LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH) {
```

Comment:
`if` are very expensive on GPUs, considering that algorithm is immutable for the kernel, can we make it a template parameter and turn this one into `if(constexpr`? (Not sure if Metal supports those)